### PR TITLE
git: implement merge/checkout via pygit2

### PR DESCRIPTION
**Pull request by pmrowla**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Finishes moving the remaining direct `gitpython` calls from experiments into `scm` and implements them in both `gitpython` and `pygit2` backends.

Related to #2215

### CODE REVIEW

File: dvc/scm/git/backend/pygit2.py [hunk 1]
```diff
@@ -234,3 +234,39 @@ def status(
...
+
+    def merge(
+        self,
+        rev: str,
+        commit: bool = True,
+        msg: Optional[str] = None,
+        squash: bool = False,
+    ) -> Optional[str]:
+        # pylint: disable=no-name-in-module
+        from pygit2 import GIT_RESET_MIXED, GitError
```
Comment (pmrowla) [inline dvc/scm/git/backend/pygit2.py:246]: pylint doesn't detect module level C imports, not sure if this is normal for pylint or if it's specific to  the way pygit2 builds/includes libgit2 exports?

**Comment by efiop** [comment]
@pmrowla Please rebase.

File: dvc/scm/git/backend/pygit2.py (hunk 1 above)
Comment (skshetry) [reply dvc/scm/git/backend/pygit2.py:246]: You can add the pygit2 here in the config:
https://github.com/iterative/dvc/blob/a2c91d20581df2e402a020a5887a72fdee7bc73a/.pylintrc#L6

File: dvc/scm/git/backend/pygit2.py [hunk 2]
```diff
@@ -84,7 +84,7 @@ def is_sha(rev: str) -> bool:
     def dir(self) -> str:
         raise NotImplementedError
 
-    def add(self, paths: Iterable[str]):
+    def add(self, paths: Iterable[str], update=False):
```
Comment (pmrowla) [inline dvc/scm/git/backend/pygit2.py:87]: `add --update` is available in `libgit2`, but `pygit2` does not import the update related index functions. I'll submit the pygit PR for this at some point in the near future.

**Review by efiop — APPROVED** [review]

**Pull request merged**